### DISCUSSION TITLE
How to apply KHR_materials_variants extension in a model?

### DISCUSSION THREAD

**Discussion by nshutijean** [google/model-viewer#2053] [Q&A]
After exporting the ```SheeChair.glb``` model in docs to ```gltf```, i found a ```KHR_materials_variants``` extension under the ```extras``` object in some nodes. 
```json
{
  "name": "SheenChair_fabric",
  "extras": {
    "gltfExtensions": {
      "KHR_materials_variants": {
        "mappings": [
          { "material": 0, "variants": [0] },
          { "material": 4, "variants": [1] }
        ]
      }
    },
    "name": "SheenChair_fabric"
  },
  "mesh": 3
 },
```
I tried to replicate it in mine, but the model seems to crash after hard-coding those lines. I tried to match the nodes I want to change with materials' numbers at the end of the file.
```json
"name": "Yellow",
  "extras": { 
      "gltfExtensions": {
          "KHR_materials_variants": {
              "mappings": [
                  {"material: 0, "variants": [0]},
              ]
          }
      }
    },
  "children": [2, 3]
},
```
Is this the best practice? Is there a better way to apply the extension to models?
I uploaded the ```html``` file with both models(glb and gltf) here: [https://repl.it/join/qekgxydc-nshutijean](url)

**Comment by elalish** [comment]
Hand-editing glTF is not super user-friendly, but if you do, I'd recommend trying the glTF VSCode extension. I imagine you probably just have a syntax error somewhere. Here's the spec: https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_variants/README.md

Also, try loading it our editor, as that should check for syntax errors (though I'm not sure how many extensions it validates currently). https://modelviewer.dev/editor/

**Reply by tom-deadline** [reply]
@elalish  is there any way to apply KHR_materials_variants to models during exporting from the studio?

**Reply by elalish** [reply]
Not yet, but we're working on it.

**Reply by nshutijean** [reply]
i've not tried this yet, but @takahirox  made a Blender plugin for this case. You might try it and see.

[https://github.com/takahirox/glTF-Blender-IO-materials-variants](https://github.com/takahirox/glTF-Blender-IO-materials-variants)

**Reply by andygomez3dart** [reply]
I have been looking for this for weeks... I will give it a try and see how it goes Many 
thanks for the info

**Reply by andygomez3dart** [reply]
It works like a charm great job !! @takahirox 
Here's a silly variation I made...
![image](https://user-images.githubusercontent.com/79045928/116486318-3c41e400-a885-11eb-8c27-2af2a3033ec4.png)

**Reply by nshutijean** [reply]
I couldn't be happier😁 I just tried the addon, and it works perfectly. Variants are now smooth to work with.
Here is a simple example I built to test it: https://gltf-variants-example.nshutijean.repl.co (model included)

Thanks @takahirox

**Reply by nshutijean** [reply]
> I have been looking for this for weeks... I will give it a try and see how it goes Many
> thanks for the info

Thanks to you too 😊
